_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+being 50% smaller! It looks like the compiler has optimized that approach better by using a lookup table
+instead of branching.
+
+#### New Feature: visualize jumps
+
+Binutils is constantly adding new features. While writing this post, I discovered a new option in
+`objdump`, `--visualize-jumps`, that was just released in binutils 2.34. This generates
+ascii art showing the branching flow within individual functions which will be very
+helpful for following control flow when stepping through assembly:
+
+```
+$ binutils/objdump  --visualize-jumps=color --disassemble=test_enum_to_str_v2
+   nrf52_example/build/nrf52.elf
+```
+
+![](img/best-of-binutils/visualize-jumps-binutils.png)
+
+### strip
+
+As the name implies, `strip` removes sections from an ELF. Most commonly you
+will see it used for stripping things like debug info from an ELF:
+
+```
+# Confirm ELF originally has debug info
+$ arm-none-eabi-readelf -S nrf52_example/build/nrf52.elf | grep debug
+  [ 8] .debug_info       PROGBITS        00000000 0200f4 00980b 00      0   0  1
+  [ 9] .debug_abbrev     PROGBITS        00000000 0298ff 001adf 00      0   0  1
+  [10] .debug_loc        PROGBITS        00000000 02b3de 003bf7 00      0   0  1
+  [11] .debug_aranges    PROGBITS        00000000 02efd5 0004f0 00      0   0  1
+  [12] .debug_ranges     PROGBITS        00000000 02f4c5 0007d8 00      0   0  1
+  [13] .debug_macro      PROGBITS        00000000 02fc9d 0031dd 00      0   0  1
+  [14] .debug_line       PROGBITS        00000000 032e7a 005447 00      0   0  1
+  [15] .debug_str        PROGBITS        00000000 0382c1 00dedb 01  MS  0   0  1
+  [18] .debug_frame      PROGBITS        00000000 046248 000d8c 00      0   0  4
+
+# Remove debug info
+$ arm-none-eabi-strip -g nrf52_example/build/nrf52.elf
+
+# Confirm ELF no longer has debug info
+$ arm-none-eabi-readelf -S nrf52_example/build/nrf52.elf | grep debug
+
+```
+
+### objcopy
+
+`objcopy` is most commonly used as the last stage of a build to generate the `.bin` or `.hex` file which
+can be flashed onto a device. Unlike an operating system where there is a "dynamic loader" that
+resolves addresses and loads libaries at runtime, on embedded devices all of the memory locations
+are typically resolved at compilation time.
+
+```bash
+$ arm-none-eabi-objcopy firmware.elf firmware.bin -O binary
+```
+
+> NOTE: For an example of leveraging using objcopy to build a bootloader for an embedded device,
+> check out [this post]({% post_url 2019-08-13-how-to-write-a-bootloader-from-scratch %}#putting-it-all-together).
+
+#### Converting binaries into an object file
+
+Sometimes for embedded development you may want to include binary artifacts into a firmware as a
+byte array. For example, sometimes a firmware will include binaries that need to be loaded onto peripherals (i.e a GPS,
+or accelerometer peripheral image). One neat thing you can do with objcopy is convert any binary blob into an object file that can
+then be linked in your image.
+
```

Comment:
Sometimes, a firmware will include binaries that need to be loaded onto peripherals, such as a GPS or accelerometer firmware image. One neat thing you can do with objcopy is convert any binary blob into an object file that can
then be linked in your image.